Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,266 @@
+    store_gradients : bool, default: `False`
+        Flag indicating whether to store an update a list of gradients for each function :math:`f_i` or just to store the snapshot point :math:` \tilde{x}` and its gradient :math:`\nabla \sum_{i=0}^{n-1}f_i(\tilde{x})`.
+
+    
+    """
+
+    def __init__(self, functions, sampler=None, update_frequency=None, store_gradients=False):
+        super(SVRGFunction, self).__init__(functions, sampler)
+
+        #  update_frequency for SVRG
+        self.update_frequency = update_frequency
+    
+        if self.update_frequency is None:
+            self.update_frequency = 2*self.num_functions
+        self.store_gradients = store_gradients
+
+        self._svrg_iter_number = 0
+        
+        self._full_gradient_at_snapshot = None
+        if self.store_gradients:
+            self._list_stored_gradients = None
+        
+        self.snapshot = None 
+
+    def gradient(self, x, out=None):
+        """ Selects a random function using the `sampler` and then calls the approximate gradient at :code:`x` or calculates a full gradient depending on the update frequency
+
+        Parameters
+        ----------
+        x : DataContainer (e.g. ImageData object)
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer (e.g. ImageData object)
+            the value of the approximate gradient of the sum function at :code:`x` 
+        """
+
+
+    
+        #  For SVRG, every `update_frequency` a full gradient step is calculated, else an approximate gradient is taken. 
+        if (np.isinf(self.update_frequency) == False and (self._svrg_iter_number % (self.update_frequency)) == 0):
+
+            return self._update_full_gradient_and_return(x, out=out)
+
+        else:
+
+            self.function_num = self.sampler.next()
+
+            
+            
+            if self.function_num >= self.num_functions or self.function_num<0 :
+                raise IndexError('The sampler has outputted an index larger than the number of functions to sample from. Please ensure your sampler samples from {0,1,...,len(functions)-1} only.')
+            return self.approximate_gradient(x, self.function_num, out=out)
+
+        
+
+    def approximate_gradient(self, x, function_num, out=None):
+        """ Calculates the stochastic gradient at the point :math:`x` by using the gradient of the selected function, indexed by :math:`i_k`, the `function_number` in {0,...,len(functions)-1}, and the full gradient at the snapshot :math:`\tilde{x}`
+            .. math ::
+                n*\nabla f_{i_k}(x_k) - n*\nabla f_{i_k}(\tilde{x}) + \nabla \sum_{i=0}^{n-1}f_i(\tilde{x})
+        
+        Note that compared with the literature, we multiply by :math:`n`, the number of functions, so that we return an approximate gradient of the whole sum function and not an average gradient.
```

Comment:
As in SAG/SAGA, you can add a Note (doc)

Reply:
Done